BF: if early attempt to parse resulted in error, setup subparsers

yarikoptic opened the pull request:
Without that using a short option for a command which is also
present on top level results in early attempt error and then no
parsers even tried, thus giving that error from the main/top parser to the user

Fixes #6814

Regression was introduced in 0b4937de63c44509001cd9c64bbeb7b24176685c ( AKA 0.16.0^2~68^2~17 )

I have not investigated possible impact on completion etc, but when tried - still worked fine.

yarikoptic commented:
wow this hit some nerve:
```
FAILED ../datalad/cli/tests/test_main.py::test_script_shims[datalad] - datala...
FAILED ../datalad/cli/tests/test_main.py::test_script_shims[git-annex-remote-datalad-archives]
FAILED ../datalad/cli/tests/test_main.py::test_script_shims[git-annex-remote-datalad]
FAILED ../datalad/cli/tests/test_main.py::test_cfg_override - FileNotFoundErr...
FAILED ../datalad/cli/tests/test_main.py::test_librarymode - datalad.runner.e...
FAILED ../datalad/cli/tests/test_main.py::test_completion - datalad.runner.ex...
FAILED ../datalad/core/distributed/tests/test_clone.py::test_invalid_args - d...
FAILED ../datalad/core/distributed/tests/test_clone.py::test_clone_crcns - Fi...
FAILED ../datalad/core/distributed/tests/test_clone.py::test_clone_datasets_root
FAILED ../datalad/core/distributed/tests/test_clone.py::test_clone_simple_local
FAILED ../datalad/core/distributed/tests/test_clone.py::test_clone_dataset_from_just_source
FAILED ../datalad/core/distributed/tests/test_clone.py::test_clone_isnot_recursive
FAILED ../datalad/core/distributed/tests/test_clone.py::test_clone_into_dataset
FAILED ../datalad/core/distributed/tests/test_clone.py::test_notclone_known_subdataset
FAILED ../datalad/core/distributed/tests/test_clone.py::test_clone_missing_commit
FAILED ../datalad/core/distributed/tests/test_clone.py::test_reckless[True]
FAILED ../datalad/core/distributed/tests/test_clone.py::test_reckless[False]
FAILED ../datalad/core/distributed/tests/test_clone.py::test_install_source_relpath
```
I guess I or someone else would need to look into it in greater depth :-/

mih commented:
I triggered a rerun of the failed appveyor runs to get a sense of whether the widespread errors are a consequence of some service failure.

Re fix: Superficially, this makes sense. Even though I did the last RF of this code, it is hard for me to say "OK" with confidence. The last RF was focused on making things faster, but it certainly did not aim for (or achieve) an improvement of this code's comprehensibility.

yarikoptic commented:
gy gy - it was the new test I added. I didn't realize that `run_main` does it in the process and doesn't run external `datalad`, so that `-C` I added was chdir'ing test process and causing all this turmoil.  I rewrote and force pushed (twice ;-) ).  Should be all good/only relevant fails now ;)

yarikoptic commented:
the only fail travis is a known https://github.com/datalad/datalad/issues/4496#issuecomment-993199936 .  As @mih gives some support in the fix ("makes sense") , we have a unittest for it now, and it seems nothing breaks -- I will proceed with merge since unlikely we would get a better fix (if there is one possible) any time soon .
